libnest/src/config/mod.rs

```diff
@@ -0,0 +1,130 @@
+//! Nest configuration parsing and handle.
+//!
+//! There are two way sto configure operations using the Nest package manager:
+//! globally (using the configuration file), or locally (through command line arguments).
+//!
+//! Within the `libnest`, many functions take a `&Config` as argument.
+//! The main reason is to allow local options to be used only for one operation, even in an
+//! asynchronous context.
+//!
+//! This module provides a `Config` structure that holds all configuration options. This includes,
+//! for example, proxy settings, cache path, mirrors etc.
+//!
+//! It also provides a way to load a `Config` from a TOML file.
+
+mod paths;
+mod repository;
+
+pub use self::paths::ConfigPaths;
+pub use self::repository::RepositoryConfig;
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+
+use failure::{Error, ResultExt};
+use lazy_static::lazy_static;
+use serde_derive::{Deserialize, Serialize};
+use toml;
+
+lazy_static! {
+    static ref NEST_PATH_CONFIG: &'static Path = Path::new("/etc/nest/config.toml");
+}
+
+/// A handle to represent a configuration for Nest.
+///
+/// This handle is given as parameter to each libnest function so they can use a custom configuration even in an asynchronous context.
+///
+/// Configuration includes proxy settings, cache path, repositories and their mirrors etc.
+///
+/// # Examples
+///
+/// ```no_run
+/// # extern crate libnest;
+/// # extern crate failure;
+/// # fn main() -> Result<(), failure::Error> {
+/// use libnest::config::Config;
+///
+/// let config = Config::load()?;
+/// # Ok(()) }
+/// ```
+#[derive(Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub struct Config {
+    #[serde(default)]
+    paths: ConfigPaths,
+    #[serde(default)]
+    repositories: HashMap<String, RepositoryConfig>,
+}
+
+impl Config {
+    /// Loads the configuration located at the default path
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # extern crate libnest;
+    /// # extern crate failure;
+    /// # fn main() -> Result<(), failure::Error> {
+    /// use libnest::config::Config;
+    ///
+    /// let config = Config::load()?;
+    /// # Ok(()) }
+    /// ```
+    #[inline]
+    pub fn load() -> Result<Config, Error> {
```

Comment:
When writing a library, you ideally do *not* want to return the `Error` type. It makes it difficult for the underlying binary to match on the original reason behind the error.

Making an enum of the different possible errors (`enum LoadConfigError`) or using the crate `error_chain` instead of `failure`  might be a better idea. I'd recommend the later one.